### DISCUSSION TITLE
Use JSONB as alias for JSON

### DISCUSSION THREAD

**Discussion by eveith** [sqlalchemy/sqlalchemy#8160] [Usage Questions] [answered]
(Crosspost from https://dba.stackexchange.com/questions/313620/use-jsonb-as-default-alias-f%c3%bcr-json)

My application uses the SQLalchemy ORM code exclusively to define the database schema. For the database, two uses cases exist. For the sake of simplicity, I'll call the first the "simple" use case, the second the "complex" use case.

Both use cases go through my application. Both use cases generate JSON data. However, the use cases differ on how the JSON data is queried later on; consequently, the reports that run on the database make use of the same application/library code, but construct different queries on the JSON attributes of the schema.

Now, the simple use case employs exclusively SQLite, while the complex use case relies on PostgreSQL exclusively. I would like to use JSONB on PostgreSQL, because all reports that run against the PostgreSQL database cast all JSON fields to JSONB. SQLite, however, does not have JSONB, obviously. Still, I would like to use the same ORM code in both cases.

How can I make SQLalchemy use JSONB when my code connects to a PostgreSQL database, but JSON in all other cases? I.e., can I change the facade of JSON for the PostgreSQL dialect?

I haven't found anything in the documentation, most probably because I don't know what to look for. The search terms I tried ("type alias", or even giving the full question) didn't turn up something useful. I have looked at the source code; each dialect/driver defines a `colspec`. I'm a bit reluctant to change that, though.

I'd appreciate any guidance. Thanks a lot in advance!

**Comment by zzzeek** [comment, marked as answer]
hi -

use [with_variant](https://docs.sqlalchemy.org/en/14/core/type_api.html#sqlalchemy.types.TypeEngine.with_variant)

    Column('data', JSON().with_variant(JSONB(), "postgresql"))

**Reply by eveith** [reply]
Oh my, its a base functionality… I was looking at all the wrong places. Thank you so much!

**Reply by zzzeek** [reply]
no worries, it's not well documented
